test/erlzk_test.erl

```diff
@@ -400,94 +431,181 @@ watch({ServerList, Timeout, _Chroot, _AuthData}) ->
+    %% connection process
+    {ok, Pid} = erlzk:connect(ServerList, Timeout, [{monitor, self()}]),
+    {Host, Port} = receive
+                       {connected, H, P} -> {H, P}
+                   after
+                       Timeout ->  ?assert(false)
+                   end,
+
+    ExistCreateWatch = ?spawn_watch({node_created, <<"/a">>}),
+    ?assertMatch({error, no_node}, erlzk:exists(Pid, "/a", ExistCreateWatch)),
+
+    %% To ensure the connection will be established towards a
+    %% different ZK host, remove the current server from the state of
+    %% `Pid'
+    sys:replace_state(Pid,
+                      fun (State) ->
+                              OldServers = element(2, State),
+                              NewServers = lists:delete({Host, Port}, OldServers),
+                              setelement(2, State, NewServers)
+                      end,
+                      infinity),
+    erlzk:kill_connection(Pid),
+
+    receive
+        DisconnectedMsg -> ?assertEqual({disconnected, Host, Port}, DisconnectedMsg)
+    after
+        Timeout -> ?assert(false)
+    end,
+    receive
+        ReconnectedMsg -> ?assertMatch({connected, NewHost, NewPort} when NewHost =/= Host orelse NewPort =/= Port,
+                                       ReconnectedMsg)
+    after
+        Timeout -> ?assert(false)
+    end,
+
+    ?assertMatch({ok, "/a"}, erlzk:create(Pid, "/a")),
+    ?assertCompleted(ExistCreateWatch),
+    ?assertMatch(ok, erlzk:delete(Pid, "/a")),
+
+    erlzk:close(Pid),
+    ok.
+
+reconnect_with_stale_watches({ServerList, Timeout, _Chroot, _AuthData}) ->
+    {ok, Pid} = connect_and_wait(ServerList, Timeout),
+
+    ?assertEqual({ok, "/a"}, erlzk:create(Pid, "/a")),
+    StaleDataWatch = ?spawn_watch({node_deleted, <<"/a">>}),
+    ?assertMatch({ok, {<<>>, _}}, erlzk:get_data(Pid, "/a", StaleDataWatch)),
+    StaleChildWatch = ?spawn_watch({node_deleted, <<"/a">>}),
+    ?assertMatch({ok, []}, erlzk:get_children(Pid, "/a", StaleChildWatch)),
+    StaleExistsWatch = ?spawn_watch({node_created, <<"/b">>}),
+    ?assertEqual({error, no_node}, erlzk:exists(Pid, "/b", StaleExistsWatch)),
+
+    ?assertEqual({ok, "/c"}, erlzk:create(Pid, "/c")),
+    ValidDataWatch = ?spawn_watch({node_data_changed, <<"/c">>}),
+    ?assertMatch({ok, {<<>>, _}}, erlzk:get_data(Pid, "/c", ValidDataWatch)),
+
+    {ok, P} = connect_and_wait(ServerList, Timeout),
+    erlzk:block_incoming_data(Pid),
+    ?assertEqual(ok, erlzk:delete(P, "/a")),
+    ?assertEqual({ok, "/b"}, erlzk:create(P, "/b")),
+    ?assertEqual(ok, erlzk:delete(P, "/b")),
+    ?assertMatch({ok, _}, erlzk:set_data(P, "/c", <<"changed">>)),
+
+    erlzk:kill_connection(Pid),
+    receive
+        {connected, _Host, _Port} -> ok
+    after
+        Timeout -> ?assert(false)
+    end,
+
+    ?assertCompleted(ValidDataWatch),
+    ?assertCompleted(StaleDataWatch),
+    ?assertCompleted(StaleChildWatch),
+
+    % This seems to be a ZooKeeper bug: the exists watch is not
```

Comment:
From zookeeper doc:
>  If /znode1 changes again, no watch event will be sent unless the client has done another read that sets a new watch.

Reply:
You're quoting the _One-time trigger_ property of watches. I don't expect this watch to trigger twice (first for the creation then the deletion of the `/b` node), the problem is that it wouldn't trigger at all.

My point here is:
* Assume there are two ZK clients, both owning the same watches at some point in time (say, zxid 1000).
* Then one of the clients disconnects.
* Then a series of changes are made to ZK.
* Then the disconnected client reconnects and resets its watches with a [SetWatches](https://github.com/apache/zookeeper/blob/12efc9f88705796cb905b46159e1d519323b1d9c/src/zookeeper.jute#L70-L75) call, providing `relativeZxid=1000`.
* Then if at this point we compare the two clients, we shall see that they have received the same events. Either a watch triggered for both of them, or none of them. Otherwise ZK does not respect `relativeZxid` properly.

In this example, this invariant is violated. The exists watch should have triggered after it was created, but the client never received this event.

Reply:
ok. understood.